### PR TITLE
CON-1861  Push Testcase results to Dashboard - volume test

### PR DESCRIPTION
**Pull request by prasannakumari-sundaraeasan**
Included changes for pushing tescase results to dashboard

### CODE REVIEW

File: test/dashboard.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+func loginDashboard() (*http.Client) {
+		
+	sessionCookie  = nil			
+	jar, _ := cookiejar.New(nil)
+	
+	client := &http.Client{
+		Jar: jar,
+	}
+	
+	var jsonStr = []byte(`{"username":"tester","password":"nimbledcstester"}`)
```
Comment (vidhutsingh) [inline test/dashboard.go]: Remove actual Array IP address and credentials with dummy string since this is public repo.

File: test/dashboard.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+import (
+	"net/http"	
+	"bytes"
+	"net/http/cookiejar"
+	"fmt"	
+	"io/ioutil"
+	"encoding/json"
+)
+
+var sessionCookie []*http.Cookie
```
Comment (vidhutsingh) [inline test/dashboard.go]: It is not good practice to use global variables. you can write a wrapper function to return these values if they are used at multiple places

File: test/dashboard.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+		"\"JenkinsJob\": \"\""+
+		"}}"
+	fmt.Println(json_string)
+	var testData = []byte(json_string)	
+	req, _:= http.NewRequest("POST", "http://10.18.223.131:4000/testcasevalues", bytes.NewBuffer(testData))
+	
+	req.Header.Add("Content-Type", "application/json")	
+	resp, err:= postDataClient.Do(req)
+	
+	if err != nil {
```
Comment (vidhutsingh) [inline test/dashboard.go:68]: Please refactor the code , use if else instead of using multple if statements

File: test/dashboard.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+		"\"RingId\": \"2\","+
+		"\"EndTimeStamp\": \"120103\","+
+		"\"TargetScope\": \"volume\","+
+		"\"Protocol\": \"iscsi\","+
+		"\"SuiteName\": \"Golang SDK\","+
+		"\"JenkinsJob\": \"\""+
+		"}}"
+	fmt.Println(json_string)
+	var testData = []byte(json_string)	
+	req, _:= http.NewRequest("POST", "http://10.18.223.131:4000/testcasevalues", bytes.NewBuffer(testData))
```
Comment (vidhutsingh) [inline test/dashboard.go]: Pls dont use actual ip address

File: test/dashboard.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
+package test
```
Comment (vidhutsingh) [inline test/dashboard.go:3]: Add copyright

File: test/dashboard.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+		"\"TargetScope\": \"volume\","+
+		"\"Protocol\": \"iscsi\","+
+		"\"SuiteName\": \"Golang SDK\","+
+		"\"JenkinsJob\": \"\","+
+		"\"RunId\": \""+runId+"\""+
+		"}}"
+	fmt.Println(json_string)
+	var testData = []byte(json_string)
+	
+	req, _:= http.NewRequest("POST", "http://10.18.223.131:4000/testcasevalues", bytes.NewBuffer(testData))
```
Comment (vidhutsingh) [inline test/dashboard.go]: Ip address ?

File: test/volume_test.go [hunk 7]
```diff
@@ -22,38 +25,61 @@ type VolumeWorkflowSuite struct {
 	volcollService      *service.VolumeCollectionService
 	initiatorGrpService *service.InitiatorGroupService
 }
-
+func (suite *VolumeWorkflowSuite) SetupTest() {
+    fmt.Println("SetupTest")
+	if testResult == "PASS" && result == "PASS" {
+	testResult = "PASS"
```
Comment (rkumpf) [inline test/volume_test.go]: This statement has no effect.  Same thing at line 58.

File: test/volume_test.go [hunk 8]
```diff
@@ -22,38 +25,61 @@ type VolumeWorkflowSuite struct {
...
 
+
 func (suite *VolumeWorkflowSuite) TearDownSuite() {
+	
+	if testResult == "PASS" && result == "PASS" {
+		testResult = "PASS"	
+	} else {
+			testResult = "FAIL"
+	}
+	fmt.Println("teardo testResult")
```
Comment (rkumpf) [inline test/volume_test.go]: *tear down?

File: test/volume_test.go [hunk 9]
```diff
@@ -22,38 +25,61 @@ type VolumeWorkflowSuite struct {
...
+	fmt.Println(testResult)
+	if pushResultToDB == true {
+	pushResultToDashboard(testResult,"C545072","Volume workflow")
+	}
+	fmt.Printf("tear down up suite\n")
```
Comment (rkumpf) [inline test/volume_test.go]: Remove "up"?

File: test/volume_test.go [hunk 10]
```diff
@@ -3,16 +3,19 @@
...
-	"github.com/stretchr/testify/assert"
-	"github.com/stretchr/testify/suite"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"	
 	"testing"
 )
 
 const volumeName = "VolumeTest"
+var testResult = "PASS"
+var result ="PASS"
```
Comment (rkumpf) [inline test/volume_test.go]: What is the difference between result and testResult?  More meaningful names might help.  Perhaps this variable could be changed to "testCaseCompleted bool".  Then the somewhat confusing 'result = "FAIL"' statement at the beginning of each test case could be changed to 'testCaseCompleted = false".

File: test/volume_test.go [hunk 11]
```diff
@@ -22,38 +25,61 @@ type VolumeWorkflowSuite struct {
...
+	
+	if testResult == "PASS" && result == "PASS" {
+		testResult = "PASS"	
+	} else {
+			testResult = "FAIL"
+	}
+	fmt.Println("teardo testResult")
+	fmt.Println(testResult)
+	if pushResultToDB == true {
+	pushResultToDashboard(testResult,"C545072","Volume workflow")
```
Comment (rkumpf) [inline test/volume_test.go]: Use "go fmt" to correct formatting problems.

File: test/volume_test.go [hunk 12]
```diff
@@ -22,38 +25,61 @@ type VolumeWorkflowSuite struct {
 	volcollService      *service.VolumeCollectionService
 	initiatorGrpService *service.InitiatorGroupService
 }
-
+func (suite *VolumeWorkflowSuite) SetupTest() {
+    fmt.Println("SetupTest")
+	if testResult == "PASS" && result == "PASS" {
+	testResult = "PASS"	
+	} else {
+		testResult = "FAIL"
```
Comment (rkumpf) [inline test/volume_test.go]: Why do we set a test result in the setup function?  Doesn't it belong in a function that runs after the test case is executed?  See https://pkg.go.dev/github.com/stretchr/testify/suite#AfterTest.

File: test/volume_test.go [hunk 13]
```diff
@@ -22,38 +25,61 @@ type VolumeWorkflowSuite struct {
...
+	
 }
 
+
 func (suite *VolumeWorkflowSuite) TearDownSuite() {
+	
+	if testResult == "PASS" && result == "PASS" {
+		testResult = "PASS"	
+	} else {
+			testResult = "FAIL"
```
Comment (rkumpf) [inline test/volume_test.go]: This redundant code could be eliminated if the SetupTest logic is moved to an AfterTest function.

File: test/common.go [hunk 14]
```diff
@@ -31,6 +31,8 @@ var sourceArrayUsername = flag.String("sourceArrayUsername", "xxx", "Source arra
 
 var sourceArrayPassword = flag.String("sourceArrayPassword", "xxx", "Source array password")
 
+var pushResultToDB = true
```
Comment (rkumpf) [inline test/common.go]: DB means dashboard?  If so, it needs to be spelled out.

File: test/dashboard.go [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+	var jsonStr = []byte(`{"username":"tester","password":"nimbledcstester"}`)	
+	req, _ := http.NewRequest("POST", "http://10.18.223.131:4000/apilogin", bytes.NewBuffer(jsonStr))
+	
+	req.Header.Add("Content-Type", "application/json")
+	fmt.Println(req)
+	resp, err := client.Do(req)
+	fmt.Println(resp)
+	
+	if err != nil {
+		panic(nil)
```
Comment (rkumpf) [inline test/dashboard.go]: Should be panic(err)?

File: test/dashboard.go [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+	
+	client := &http.Client{
+		Jar: jar,
+	}
+	
+	var jsonStr = []byte(`{"username":"tester","password":"nimbledcstester"}`)	
+	req, _ := http.NewRequest("POST", "http://10.18.223.131:4000/apilogin", bytes.NewBuffer(jsonStr))
+	
+	req.Header.Add("Content-Type", "application/json")
+	fmt.Println(req)
```
Comment (rkumpf) [inline test/dashboard.go]: Does this print credentials?  That should be avoided.

File: test/dashboard.go [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+	resp, err := client.Do(req)
+	fmt.Println(resp)
+	
+	if err != nil {
+		panic(nil)
+	}
+	if err == nil{
```
Comment (rkumpf) [inline test/dashboard.go]: This condition can be removed.  It will never be false.

File: test/dashboard.go (hunk 2 above)
Comment (prasannakumari-sundaraeasan) [reply test/dashboard.go]: Hi Vidhut,
Since we are maintaining same session and have same runid for all testsuites, I am using global variables.

**Review by vidhutsingh — COMMENTED** [review]
Please change copyright 2021 to new files 
e.g
Copyright 2021 Hewlett Packard Enterprise Development LP

**Comment by vidhutsingh** [comment]
@prasannakumari-sundaraeasan , please address rkumph review comments as well. thanks

**Comment by vidhutsingh** [comment]
There is DCO error, please use git commit -s -m "" while making code commit to avoid DCS errors .

File: test/dashboard.go (hunk 17 above)
Comment (prasannakumari-sundaraeasan) [reply test/dashboard.go]: removed this condition

File: test/dashboard.go (hunk 16 above)
Comment (prasannakumari-sundaraeasan) [reply test/dashboard.go]: removed println

File: test/dashboard.go (hunk 15 above)
Comment (prasannakumari-sundaraeasan) [reply test/dashboard.go]: changed the code

File: test/common.go (hunk 14 above)
Comment (prasannakumari-sundaraeasan) [reply test/common.go]: changed to meaningful name

File: test/volume_test.go (hunk 13 above)
Comment (prasannakumari-sundaraeasan) [reply test/volume_test.go]: Refactored the code

File: test/volume_test.go (hunk 10 above)
Comment (prasannakumari-sundaraeasan) [reply test/volume_test.go]: refactored the code

File: test/volume_test.go (hunk 7 above)
Comment (prasannakumari-sundaraeasan) [reply test/volume_test.go]: refactored the code

File: test/volume_test.go (hunk 8 above)
Comment (prasannakumari-sundaraeasan) [reply test/volume_test.go]: corrected spelling

File: test/volume_test.go (hunk 9 above)
Comment (prasannakumari-sundaraeasan) [reply test/volume_test.go]: Removed "up"

File: test/volume_test.go (hunk 11 above)
Comment (prasannakumari-sundaraeasan) [reply test/volume_test.go]: fixed formatting problem

**Pull request closed without merging**